### PR TITLE
Create a common directory for quantization code and functionality.

### PR DESCRIPTION
**Pull request by nkreeger**
This change creates a new folder for shared quantization code in `onnxruntime/core/quantization`. Going forward, all quantized kernels and tests should reference this canonical set of functions for using and accelerating quantized operations. This first change introduces a global `quantization.h` file with utility functions to handle quantizing and dequantizing vectors and lists of values to `int8_t` and `uint8_t`.

All code in this new header matches the logic used in the python quantization toolkit. Keeping these two aligned will ensure that quantization tests in c++ and python should match the exact implementation.

Additionally, I'm introducing a new class to hold scale/zp values - `quantization::Params<T>`. This class should be used to ensure that kernel or tests do not use an unsupported quantized integer width.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc [hunk 1]
```diff
@@ -64,22 +60,22 @@ Status ComputeInternal(OpKernelContext* context, float epsilon) {
...
-      has_segment_embedding ? *(segment_embedding_scale->template Data<float>()) : 0.0f;
-  T2 segment_embedding_zero_point_data =
-      has_segment_embedding ? *(segment_embedding_zero_point->template Data<T2>()) : 0;
-
-  float gamma_scale_data = *(gamma_scale->template Data<float>());
-  T2 gamma_zero_point_data = *(gamma_zero_point->template Data<T2>());
-
-  float beta_scale_data = *(beta_scale->template Data<float>());
-  T2 beta_zero_point_data = *(beta_zero_point->template Data<T2>());
+  quantization::Params<T2> word_embedding_params =
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc:63]: >T2 [](http://example.com/codeflow?start=23&length=2)

nit: could we use a more meaningful type name

File: onnxruntime/core/quantization/quantization.h [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  if constexpr (std::is_same<T, int8_t>::value) {
+    T_min = -127;
+  }
+
+  float min = std::numeric_limits<float>::max();
+  float max = std::numeric_limits<float>::min();
+  for (size_t i = 0; i < size; ++i) {
+    min = std::min(min, data[i]);
+    max = std::max(max, data[i]);
+  }
```
Comment (yufenglee) [inline onnxruntime/core/quantization/quantization.h:115]: use std::minmax_element?

File: onnxruntime/core/quantization/quantization.h [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+namespace quantization {
+
+#define ORT_STATIC_ASSERT_QUANTIZATION_TYPES(T)                            \
+  static_assert(                                                           \
+      !std::is_same<T, int8_t>::value || !std::is_same<T, uint8_t>::value, \
+      "Only int8_t and uint8_t are supported quantization formats.");
+
+// Basic quantization params structure.
+template <typename T>
+class Params {
```
Comment (yufenglee) [inline onnxruntime/core/quantization/quantization.h:29]: >class [](http://example.com/codeflow?start=0&length=5)

nit: struct?

File: onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc (hunk 1 above)
Comment (nkreeger) [reply onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc:63]: Sure - done. We should uniformally make a stance here and update the style guide. I see a bunch of kernels using T1/T2/T3/T/etc. I see that even stuff in `contrib_defs.cc` also use this style.

File: onnxruntime/core/quantization/quantization.h (hunk 2 above)
Comment (nkreeger) [reply onnxruntime/core/quantization/quantization.h:115]: Thought about it - but `std::minmax_element` requires an iterator - and this signature operates on a type pointer:

https://en.cppreference.com/w/cpp/algorithm/minmax_element

File: onnxruntime/core/quantization/quantization.h (hunk 3 above)
Comment (nkreeger) [reply onnxruntime/core/quantization/quantization.h:29]: Done - I went back and forth here when building this. I think `struct` makes sense if I'm going to allow access to members w/o accessors.

File: onnxruntime/core/quantization/quantization.h (hunk 2 above)
Comment (yufenglee) [reply onnxruntime/core/quantization/quantization.h:115]: pointer is a type of iterator. minmax_element(data, data+size) should work.

Comment (nkreeger) [reply onnxruntime/core/quantization/quantization.h]: Cool - done.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**